gitpod-io/gitpod
# Return and show error when stopping workspace in maintenance mode

## Description


Update `StopWorkspace` in server to await and return any error from the call to ws-manager, such that we can show an error when the stop fails due to the cluster being in maintenance mode

Both `TakeSnapshot` and `StopWorkspace` check for a failed precondition status code (=maintenance enabled), and return a more human-readable error message.

Also show a toast when stopping a workspace fails in the dashboard:
<img width="1527" alt="image" src="https://github.com/gitpod-io/gitpod/assets/9721064/75eaed08-044b-4018-80c5-0b24de7090de">

Inside a workspace using the `gp` CLI now shows the error messages returned by server:
```console
$ gp snapshot
jsonrpc2: code 412 message: Cannot take a snapshot because the workspace cluster is under maintenance. Please try again in a few minutes

$ gp stop
jsonrpc2: code 412 message: Cannot stop the workspace because the workspace cluster is under maintenance. Please try again in a few minutes
```

## Related Issue(s)

Fixes WKS-208

## How to test


Open a preview env, start a workspace, and then enable maintenance mode by creating this ConfigMap:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: ws-manager-mk2-maintenance-mode
  namespace: default
data:
  config.json: |
    {
      "enabledUntil": "2024-06-09T15:10:13Z"
    }
```

Try to take a snapshot or stop the running workspace from inside the workspace or through the dashboard, and observe the error messages

## Documentation


#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

</details>

<details>
<summary>Installer</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold
